src/Rings/MPolyMap/MPolyRing.jl

```diff
@@ -132,17 +132,122 @@ end
 #
 ################################################################################
 
+# Some additional methods needed for the test in the constructor for MPolyAnyMap
+_is_gen(x::MPolyQuoRingElem) = _is_gen(lift(x))
+_is_gen(x::MPolyDecRingElem) = is_gen(forget_grading(x))
+_is_gen(x::MPolyRingElem) = is_gen(x)
+# default method; overwrite if you want this to work for your rings.
+_is_gen(x::NCRingElem) = false
+
+# In case there is a type of ring elements for which hashing is correctly implemented 
+# and does not throw an error, this gives the opportunity to overwrite the `allunique`
+# to be used within the constructor for maps. 
+function _allunique(lst::Vector{T}) where {T<:MPolyRingElem}
+  return allunique(lst)
+end
+
+# We have a lot of rings which do/can not implement correct hashing. 
+# So we make the following the default.
+function _allunique(lst::Vector{T}) where {T<:RingElem}
+  return all(!(x in lst[i+1:end]) for (i, x) in enumerate(lst))
+end
+
+function _evaluate_plain(F::MPolyAnyMap{<:MPolyRing, <:MPolyRing}, u)
+  if isdefined(F, :variable_indices)
+    S = codomain(F)::MPolyRing
+    kk = coefficient_ring(S)
+    r = ngens(S)
+    ctx = MPolyBuildCtx(S)
+    for (c, e) in zip(AbstractAlgebra.coefficients(u), AbstractAlgebra.exponent_vectors(u))
+      ee = [0 for _ in 1:r]
+      for (i, k) in enumerate(e)
+        ee[F.variable_indices[i]] = k
+      end
+      push_term!(ctx, kk(c), ee)
+    end
+    return finish(ctx)
+  end
+
+  return evaluate(u, F.img_gens)
+end
+
 function _evaluate_plain(F::MPolyAnyMap{<: MPolyRing}, u)
   return evaluate(u, F.img_gens)
 end
 
 # See the comment in MPolyQuo.jl
 function _evaluate_plain(F::MPolyAnyMap{<:MPolyRing, <:MPolyQuoRing}, u)
+  if isdefined(F, :variable_indices)
+    S = base_ring(codomain(F))::MPolyRing
+    r = ngens(S)
+    ctx = MPolyBuildCtx(S)
+    for (c, e) in zip(coefficients(u), exponents(u))
+      ee = [0 for _ in 1:r]
+      for (i, k) in enumerate(e)
+        ee[F.variable_indices[i]] = k
+      end
+      push_term!(ctx, c, ee)
+    end
```

Comment:
Any chance we can deduplicate this part? It seems to be repeated a few times.

Should probably be something like
```
function _bla_bla(S, variable_indices, u)
...
```
?

Reply:
Probably, yes. I'll look into it.